### PR TITLE
feat: add mcopy implementation

### PR DESCRIPTION
**Pull request by charles-cooper**
### What was wrong?

- Re: EIP-5656, use memoryview slice/assign. note that this has the correct performance characteristics; see that cpython uses memmove here: https://github.com/python/cpython/blob/d2f1b0eb4956b4923f111c7c740ba7ab25f3312d/Objects/memoryobject.c#L357

### CODE REVIEW

**Comment by pacrob** [comment]
Thanks for getting this going! I've switched it to base off the existing `cancun-network-upgrade` branch, and fix the conflicts now.

**Comment by fselmo** [comment]
@charles-cooper @pacrob want to make another pass for more eyes on the changes. I got all the tests passing. Took me way too long to realize some of those state tests don't take EIP-4788 into account 😔.

File: eth/vm/logic/memory.py [hunk 1]
```diff
@@ -38,3 +44,20 @@ def mload(computation: ComputationAPI) -> None:
...
+
+
+def mcopy(computation: ComputationAPI) -> None:
+    dst, src, length = computation.stack_pop_ints(3)
+
+    # Extend the memory to cover the entire range of the copy. This will make sure that
+    # the memory is allocated for the entire range and that the gas cost is correctly
+    # calculated.
+    computation.extend_memory(src, length)
+    computation.extend_memory(dst, length)
```
Comment (charles-cooper) [inline eth/vm/logic/memory.py:56]:
```suggestion
    computation.extend_memory(max(src, dst), length)
```

**Comment by charles-cooper** [comment]
looks good, we can combine the two calls to `extend_memory()` into one

**Pull request merged**